### PR TITLE
Add analyzer template

### PR DESCRIPTION
**Pull request by YuliiaKovalova**
## Related
[#9633](https://github.com/dotnet/msbuild/issues/9633)

## Details
Add a basic content for a custom analyzer.

Given the multitude of prerequisites necessary for this operation, this template aims to simplify the process by automating the following tasks:

1. Copying third-party dependencies.
2. Invocation of intrinsic function `[MSBuild]::RegisterAnalyzer(string)` from a custom .props file.
3. Modification of the output structure for NuGet packages, shifting from the standard `/lib/platform/*.dll` to `lib/*.dll`.

### CODE REVIEW

File: template_feed/Microsoft.AnalyzerTemplate/README.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# MSBuild Custom Analyzer Template
+
+## Overview
+MSBuild Custom Analyzer Template is a .NET template designed to streamline the creation of MSBuild analyzer libraries. This template facilitates the development of custom analyzers targeting .NET Standard, enabling developers to inspect and enforce conventions, standards, or patterns within C# project files (.csproj).
```
Comment (JanKrivanek) [inline template_feed/Microsoft.AnalyzerTemplate/README.md:4]:
```suggestion
MSBuild Custom Analyzer Template is a .NET template designed to streamline the creation of MSBuild analyzer libraries. This template facilitates the development of custom analyzers targeting .NET Standard, enabling developers to inspect and enforce conventions, standards, or patterns within their MSBuild builds.
```

We do not limit to .csprojs, plus we'l be able to run from logs as well

File: template_feed/Microsoft.AnalyzerTemplate/README.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+## Features
+- Simplified template for creating MSBuild analyzer libraries.
+- Targeting .NET Standard for cross-platform compatibility.
+- Provides a starting point for implementing custom analysis rules.
+
+## Getting Started
+To use the MSBuild Custom Analyzer Template, follow these steps:
+1. Install the template using the following command:
+   ```bash
+   dotnet new -i msbuildanalyzer
```
Comment (JanKrivanek) [inline template_feed/Microsoft.AnalyzerTemplate/README.md:15]:
```suggestion
   dotnet new install msbuildanalyzer
```
`-i` is the old convention, with limited support for tab completition etc.

File: template_feed/Microsoft.AnalyzerTemplate/Company.AnalyzerTemplate.csproj [hunk 3]
```diff
@@ -0,0 +1,53 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+	  <TargetFramework>netstandard2.0</TargetFramework>
+	  <DevelopmentDependency>true</DevelopmentDependency>
+	  <IncludeBuildOutput>false</IncludeBuildOutput>
+	  <GeneratePackageOnBuild>True</GeneratePackageOnBuild>
+	  <!-- The output structure was modified for msbuild develomplent needs.-->
+    <NoWarn>NU5101;NU5128</NoWarn>
```
Comment (JanKrivanek) [inline template_feed/Microsoft.AnalyzerTemplate/Company.AnalyzerTemplate.csproj:9]: The indention is different here

**Review by JanKrivanek — COMMENTED** [review]
Great!!

Do you plan to add a reference of Microsoft.Build package and add a dummy implementation? That would be likely helpfull as well.

The version of the PackageDependency can be a parameter in template (defaulting to latest current release, https://github.com/dotnet/msbuild/blob/main/eng/Versions.props#L5); the sample implementation can be part of the Microsoft.Build, added here as link - ensuring the code is up to date. Though I haven't thought about that too deeply - if it turns complicated, feel free to merge the current version and create separate workitem for improvements

**Comment by YuliiaKovalova** [comment]
> Great!!
> 
> Do you plan to add a reference of Microsoft.Build package and add a dummy implementation? That would be likely helpfull as well.
> 
> The version of the PackageDependency can be a parameter in template (defaulting to latest current release, https://github.com/dotnet/msbuild/blob/main/eng/Versions.props#L5); the sample implementation can be part of the Microsoft.Build, added here as link - ensuring the code is up to date. Though I haven't thought about that too deeply - if it turns complicated, feel free to merge the current version and create separate workitem for improvements

Jan, I have added package reference to Microsoft.Build

File: template_feed/Microsoft.AnalyzerTemplate/.template.config/template.json [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+      {
+        "path": "Company.AnalyzerTemplate.csproj"
+      }
+   ],
+   "symbols": {
+    "MicrosoftBuildVersion": {
+      "type": "parameter",
+      "description": "Overrides the default Microsoft.Build version where analyzer's interfaces are placed",
+      "datatype": "text",
+      "defaultValue": "17.9.5",
```
Comment (JanKrivanek) [inline template_feed/Microsoft.AnalyzerTemplate/.template.config/template.json:30]: Something for the future iteration - the version here should be in sync with Version.props - we can either ensure that during pcaking, or just simply ad a manual steps to our release checklist

File: template_feed/Microsoft.AnalyzerTemplate/Analyzer1.cs [hunk 5]
```diff
@@ -0,0 +1,9 @@
+﻿using System;
+
+namespace Company.AnalyzerTemplate
+{
+    public class Analyzer1
+    {
+        // will be added later.
+    }
```
Comment (JanKrivanek) [inline template_feed/Microsoft.AnalyzerTemplate/Analyzer1.cs:8]: Let's put a skeleton of actual implementation (needs `using Microsoft.Build.Experimental.BuildCop`)

```suggestion
    public sealed class Analyzer1 : BuildAnalyzer
    {
        public static BuildAnalyzerRule SupportedRule = new BuildAnalyzerRule("X01234", "Title",
            "Description", "Category",
            "Message format: {0}",
            new BuildAnalyzerConfiguration() { Severity = BuildAnalyzerResultSeverity.Warning, IsEnabled = true });

        public override string FriendlyName => "Company.Analyzer1";

        public override IReadOnlyList<BuildAnalyzerRule> SupportedRules { get; } =[SupportedRule];

        public override void Initialize(ConfigurationContext configurationContext)
        {
            // configurationContext to be used only if analyzer needs external configuration data.
        }

        public override void RegisterActions(IBuildCopRegistrationContext registrationContext)
        {
            registrationContext.RegisterEvaluatedPropertiesAction(EvaluatedPropertiesAction);
        }
        
        private void EvaluatedPropertiesAction(BuildCopDataContext<EvaluatedPropertiesAnalysisData> context)
        {
            context.ReportResult(BuildCopResult.Create(
                SupportedRule,
                ElementLocation.EmptyLocation,
                "Argument for the message format");
        }
    }
```

**Review by JanKrivanek — COMMENTED** [review]
Do we need an explicit pack execution for this - so that actual package is produced as part of the build?

**Review by rokonec — APPROVED** [review]

File: template_feed/Microsoft.AnalyzerTemplate/.template.config/template.json (hunk 4 above)
Comment (YuliiaKovalova) [reply template_feed/Microsoft.AnalyzerTemplate/.template.config/template.json:30]: @GangWang01, could you take care of it? the first option is preferable. Feel free to create a separate work item for that.

**Review by JanKrivanek — APPROVED** [review]
Great work - thanks!!

I added https://github.com/dotnet/msbuild/issues/9915 to handle the packing

File: template_feed/Microsoft.AnalyzerTemplate/.template.config/template.json (hunk 4 above)
Comment (JanKrivanek) [reply template_feed/Microsoft.AnalyzerTemplate/.template.config/template.json:30]: @GangWang01 - would you be able to handle https://github.com/dotnet/msbuild/issues/9915 as well? Feel free to contact me with any questions

Comment (GangWang01) [reply template_feed/Microsoft.AnalyzerTemplate/.template.config/template.json:30]: Created #9923. And self assigned with both issues.

Comment (JaynieBai) [reply template_feed/Microsoft.AnalyzerTemplate/.template.config/template.json:30]: > Something for the future iteration - the version here should be in sync with Version.props - we can either ensure that during pcaking, or just simply ad a manual steps to our release checklist

@JanKrivanek   Could you give me more info how to add a manual steps to our release checklist?

Comment (YuliiaKovalova) [reply template_feed/Microsoft.AnalyzerTemplate/.template.config/template.json:30]: please include this info to this document
https://github.com/dotnet/msbuild/pull/9958

**Pull request merged**